contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+> kubectl rollback <controller-kind> <controller-name> --revision <any-state-name>
+```
+
+Here `kubectl rollback` simply applies the manifest generated by the 
+`kubectl history` command. We propose that the `--revision` argument be 
+optional, and that its exclusion indicates that the controller should be 
+reverted to its immediately prior state.
+
+## Tests
+
+1. Controllers can create an AnyState containing a revision of their target 
+Object state.
+1. Controllers can reconstruct their revision history.
+1. Controllers can't update an AnyState's `.Data`.
+1. Controllers can delete an AnyState to maintain their history with respect to 
+the configured `RevisionHistoryLimit`.
+
+## Appendix
+ 
+### Hashing
+We will require a CRHF (collision resistant hash function), but, as we expect 
+no adversaries, such a function need not be resistant to pre-image and 
+secondary pre-image attacks.
+As the property of interest is primarily collision resistance, and as we 
+provide a method of [collision resolution](#collision-resolution), both 
+cryptographically strong functions, such as Secure Hash Algorithm 2 (SHA-2), 
+and non-cryptographic functions, such as Fowler-Noll-Vo (FNV) are applicable. 
+
+### Collision Resolution
+As the function selected for hashing may not be cryptographically strong and may 
+produce collisions, we need a method for collision resolution. To demonstrate 
+its feasibility, we construct such a scheme here. However, this proposal does 
+not mandate its use.
+
+Given a hash function with output size `HashSize` defined 
+as `func H(s srtring) [HashSize] byte`, in order to resolve collisions we 
+define a new function `func H'(s string, n int) [HashSize]byte` where `H'` 
+returns the result of invoking `H` on the concatenation of `s` with the string 
+value of `n`. We define a third function 
+`func H''(s string, exists func (string) bool)(int,[HashSize]byte)`. `H''` 
+will start with `n := 0` and compute `s' := H'(s,n)`, incrementing `n` when 
+`exists(s')` returns true, until `exists(s')`  returns false. After this it will 
+return `n,s'`.
+
+For our purposes, the implementation of the `exists` function will attempt to 
+create a `.Named` AnyState via the API Server using a 
+[unique name generation](#unique-name-generation). If creation fails, due to a 
+conflict, the method returns false.
+
+### Unique Name Generation
+We can use our [hash function](#hashsing) and 
+[collision resolution](#collision-resolution) scheme to generate a system 
+wide unique identifier for an Object based on a deterministic non-unique prefix 
+and a serialized representation of the Object. Kuberentes Object's `.Names` must 
+conform to a DNS subdomain. Therefore, the total length of the unique identifier
+must not exceed 255, and in practice 253, characters. We can generate a unique 
+identifier that meets this constraint by selecting a hash function such that the 
+output length is equal to `253-len(prefix)` and applying our [hash](#hashing) 
```

Comment:
I think that Deployments should use the same concept where possible to deal with ReplicaSets, but, in this case, and in others, they will likely need special handling. Also, this is just to demonstrate that the approach is plausible, we will need to think about the naming implementation, and it will have to be based on the new hash function that will be used for deployment.